File: apps/web/lib/universe.ts
import { parseInstrumentSearch, type Instrument } from "./instruments";
import type { RecordData } from "./contracts";

export type ReviewedUniverseInstrument = Instrument & { source: string; provider: string; source_id: string; metadata_hash: string; eligibility_proof_hash: string; ethical_proof_hash: string; verified_until: string; instrument_type: "STOCK" };
export type UniversePage = { coverage: "reviewed_manifest"; mode: "live"; complete_broker_universe: false; total: number; offset: number; limit: number; evaluated_at: string; catalogue_hash: string; instruments: ReviewedUniverseInstrument[] } | { coverage: "previously_researched_only"; instruments: RecordData[] };

export function parseUniverse(value: unknown): UniversePage {
  const fail = (): never => { throw new Error("Verified ISA coverage is unavailable. Please retry."); };
  if (!value || typeof value !== "object" || Array.isArray(value)) return fail();
  const data = value as Record<string, unknown>;
  if (data.coverage === "previously_researched_only") {
    if (!Array.isArray(data.instruments) || data.instruments.length > 100 || data.instruments.some(item => !item || typeof item !== "object" || Array.isArray(item))) return fail();
    return { coverage: "previously_researched_only", instruments: data.instruments as RecordData[] };
  }
  if (data.coverage !== "reviewed_manifest" || data.mode !== "live" || data.complete_broker_universe !== false || !Number.isSafeInteger(data.offset) || Number(data.offset) < 0 || Number(data.offset) > 10000) return fail();
  const hash = (item: unknown): string => typeof item === "string" && /^[a-f0-9]{64}$/.test(item) ? item : fail();
  const timestamp = (item: unknown): string => typeof item === "string" && item.length <= 40 && /T.*(?:Z|[+-]\d{2}:\d{2})$/.test(item) && Number.isFinite(Date.parse(item)) ? item : fail();
  const text = (item: unknown): string => typeof item === "string" && item.length > 0 && item.length <= 500 && !/[\u0000-\u001f\u007f]/u.test(item) ? item : fail();
  // Reuse canonical identity validation; this directory supports a larger offset.
  const identities = parseInstrumentSearch({ ...data, coverage: "reviewed_catalogue", offset: 0 });
  const rows = data.instruments as Record<string, unknown>[];
  const evaluated = timestamp(data.evaluated_at);
  const instruments = identities.instruments.map((identity, index): ReviewedUniverseInstrument => {
    const row = rows[index], expiry = timestamp(row.verified_until);
    if (!identity.research_allowed || identity.eligibility !== "VERIFIED_ELIGIBLE" || row.instrument_type !== "STOCK" || Date.parse(expiry) <= Date.parse(identity.verified_at) || Date.parse(identity.verified_at) > Date.parse(evaluated)) return fail();
    return { ...identity, source: text(row.source), provider: text(row.provider), source_id: text(row.source_id), metadata_hash: hash(row.metadata_hash), eligibility_proof_hash: hash(row.eligibility_proof_hash), ethical_proof_hash: hash(row.ethical_proof_hash), verified_until: expiry, instrument_type: "STOCK" };
  });
  return { coverage: "reviewed_manifest", mode: "live", complete_broker_universe: false, instruments, total: identities.total, offset: Number(data.offset), limit: identities.limit, evaluated_at: evaluated, catalogue_hash: hash(data.catalogue_hash) };
}

export function currentUniverse(page: Extract<UniversePage, { mode: "live" }>, now: number): ReviewedUniverseInstrument[] {
  return page.instruments.filter(item => Date.parse(item.verified_at) <= now && Date.parse(item.verified_until) > now);
}

File: src/money/data/universe.py
"""Bounded qualified GBX stock subset, not the complete accessible broker universe.

This service consumes the separately qualified, hashed Money live manifest. It
does not assert account type or purchase availability, fetch an account, broaden
the mandate, or promote public R&D listings.
"""

from __future__ import annotations

import re
from datetime import datetime
from typing import Literal

from pydantic import AwareDatetime, Field, field_validator

from money.adapters.eligibility import ELIGIBILITY_MAXIMUM_AGE
from money.data.instruments import InstrumentCatalogue, InstrumentSearchResult, ReviewedInstrument
from money.schemas.contracts import Contract, ResearchMandate, content_hash


class StockUniverseQuery(Contract):
    query: str = Field(default="", max_length=80, pattern=r"^[^\x00-\x1f\x7f]*$")
    limit: int = Field(default=20, ge=1, le=50)
    offset: int = Field(default=0, ge=0, le=10000)

    @field_validator("query")
    @classmethod
    def normalize_query(cls, value: str) -> str:
        return value.strip()


class StockUniverseInstrument(InstrumentSearchResult):
    eligibility: Literal["VERIFIED_ELIGIBLE"] = "VERIFIED_ELIGIBLE"
    research_allowed: Literal[True] = True
    synthetic: Literal[False] = False
    currency: Literal["GBP", "GBX"]
    instrument_type: Literal["STOCK"] = "STOCK"
    source: str
    provider: str
    source_id: str
    metadata_hash: str = Field(pattern=r"^[a-f0-9]{64}$")
    eligibility_proof_hash: str = Field(pattern=r"^[a-f0-9]{64}$")
    ethical_proof_hash: str = Field(pattern=r"^[a-f0-9]{64}$")
    verified_until: AwareDatetime


class StockUniversePage(Contract):
    instruments: tuple[StockUniverseInstrument, ...]
    total: int = Field(ge=0)
    limit: int = Field(ge=1, le=50)
    offset: int = Field(ge=0, le=10000)
    coverage: Literal["reviewed_manifest"] = "reviewed_manifest"
    complete_broker_universe: Literal[False] = False
    mode: Literal["live"] = "live"
    evaluated_at: AwareDatetime
    catalogue_hash: str = Field(pattern=r"^[a-f0-9]{64}$")


class ReviewedStockUniverse:
    """Read-time eligibility checks over an immutable reviewed release catalogue.

    The caller must use the validated manifest loader and retain its commercial
    data-rights checks. This view does not grant provider or redistribution rights.
    """

    def __init__(self, catalogue: InstrumentCatalogue) -> None:
        if catalogue.mode != "live" or not catalogue.qualifications:
            raise ValueError("UNIVERSE_UNAVAILABLE")
        self._catalogue = catalogue
        self._hash = content_hash(
            {
                "entries": [
                    {
                        "metadata": item.metadata.model_dump(mode="json"),
                        "identifiers": item.identifiers.model_dump(mode="json")
                        if item.identifiers
                        else None,
                        "eligibility_proof_hash": item.eligibility_proof_hash,
                        "ethical_proof_hash": item.ethical_proof_hash,
                    }
                    for item in sorted(catalogue.entries, key=lambda entry: entry.metadata.ticker)
                ],
                "qualifications": [
                    item.model_dump(mode="json")
                    for item in sorted(catalogue.qualifications, key=lambda entry: entry.provider)
                ],
            }
        )

    @staticmethod
    def _proofs_present(item: ReviewedInstrument) -> bool:
        return all(
            isinstance(digest, str) and re.fullmatch(r"[a-f0-9]{64}", digest)
            for digest in (item.eligibility_proof_hash, item.ethical_proof_hash)
        )

    def page(
        self,
        query: StockUniverseQuery,
        *,
        mandate: ResearchMandate,
        now: datetime,
    ) -> StockUniversePage:
        """Return only current, verified, mandate-compliant reviewed instruments."""
        if now.tzinfo is None or now.utcoffset() is None:
            raise ValueError("ELIGIBILITY_TIMESTAMP_INVALID")
        self._catalogue.require_current(now)
        entries = sorted(
            (
                item
                for item in self._catalogue.entries
                if self._proofs_present(item)
                and item.identifiers is not None
                and not item.failures(mandate, now)
                and (not query.query or item.matches(query.query))
            ),
            key=lambda item: (item.metadata.ticker != query.query.upper(), item.metadata.ticker),
        )
        instruments = []
        for item in entries[query.offset : query.offset + query.limit]:
            assert item.identifiers is not None
            assert item.eligibility_proof_hash is not None and item.ethical_proof_hash is not None
            metadata = item.metadata
            # Proof admission was checked above. Never add time to the original
            # verification on cache reads, and respect earlier mapping expiry.
            expiry = min(
                metadata.verified_at + ELIGIBILITY_MAXIMUM_AGE,
                item.identifiers.valid_until,
                *(qualification.valid_until for qualification in self._catalogue.qualifications),
            )
            if metadata.ethical_clearance is not None:
                expiry = min(expiry, metadata.ethical_clearance.valid_until)
            instruments.append(
                StockUniverseInstrument(
                    **item.result(now, synthetic=False).model_dump(exclude={"instrument_type"}),
                    instrument_type="STOCK",
                    source=metadata.source,
                    provider=metadata.provider,
                    source_id=metadata.source_id,
                    metadata_hash=content_hash(metadata.model_dump(mode="json")),
                    eligibility_proof_hash=item.eligibility_proof_hash,
                    ethical_proof_hash=item.ethical_proof_hash,
                    verified_until=expiry,
                )
            )
        return StockUniversePage(
            instruments=tuple(instruments),
            total=len(entries),
            limit=query.limit,
            offset=query.offset,
            evaluated_at=now,
            catalogue_hash=self._hash,
        )


# Import compatibility only. None of these aliases asserts ISA qualification.
IsaUniverseQuery = StockUniverseQuery
IsaUniverseInstrument = StockUniverseInstrument
IsaUniversePage = StockUniversePage
ReviewedIsaUniverse = ReviewedStockUniverse
